engine/runtime-language-arrow/src/main/java/org/enso/interpreter/arrow/runtime/ArrowFixedSizeArrayBuilder.java

```diff
@@ -0,0 +1,353 @@
+    static Object doDay(ArrowFixedSizeArrayBuilder receiver, long index)
+        throws UnsupportedMessageException {
+      return ArrowFixedArrayDate.readDay(receiver.buffer, index);
+    }
+
+    @Specialization(guards = "receiver.getUnit() == Date64")
+    static Object doMilliseconds(ArrowFixedSizeArrayBuilder receiver, long index)
+        throws UnsupportedMessageException {
+      return ArrowFixedArrayDate.readMilliseconds(receiver.buffer, index);
+    }
+
+    @Specialization(guards = "receiver.getUnit() == Int8")
+    public static Object doByte(ArrowFixedSizeArrayBuilder receiver, long index)
+        throws UnsupportedMessageException, InvalidArrayIndexException {
+      var at =
+          ArrowFixedArrayInt.adjustedIndex(receiver.buffer, receiver.unit, receiver.size, index);
+      if (receiver.buffer.isNull((int) index)) {
+        return NullValue.get();
+      }
+      return receiver.buffer.get(at);
+    }
+
+    @Specialization(guards = "receiver.getUnit() == Int16")
+    public static Object doShort(ArrowFixedSizeArrayBuilder receiver, long index)
+        throws UnsupportedMessageException, InvalidArrayIndexException {
+      var at =
+          ArrowFixedArrayInt.adjustedIndex(receiver.buffer, receiver.unit, receiver.size, index);
+      if (receiver.buffer.isNull((int) index)) {
+        return NullValue.get();
+      }
+      return receiver.buffer.getShort(at);
+    }
+
+    @Specialization(guards = "receiver.getUnit() == Int32")
+    public static Object doInt(ArrowFixedSizeArrayBuilder receiver, long index)
+        throws UnsupportedMessageException, InvalidArrayIndexException {
+      var at =
+          ArrowFixedArrayInt.adjustedIndex(receiver.buffer, receiver.unit, receiver.size, index);
+      if (receiver.buffer.isNull((int) index)) {
+        return NullValue.get();
+      }
+      return receiver.buffer.getInt(at);
+    }
+
+    @Specialization(guards = "receiver.getUnit() == Int64")
+    public static Object doLong(ArrowFixedSizeArrayBuilder receiver, long index)
+        throws UnsupportedMessageException, InvalidArrayIndexException {
+      var at =
+          ArrowFixedArrayInt.adjustedIndex(receiver.buffer, receiver.unit, receiver.size, index);
+      if (receiver.buffer.isNull((int) index)) {
+        return NullValue.get();
+      }
+      return receiver.buffer.getLong(at);
+    }
+  }
+
+  @ExportMessage
+  @ImportStatic(LogicalLayout.class)
+  static class WriteArrayElement {
+
+    @Specialization(guards = "receiver.getUnit() == Date32")
+    static void doDay(
+        ArrowFixedSizeArrayBuilder receiver,
+        long index,
+        Object value,
+        @Cached.Shared("interop") @CachedLibrary(limit = "1") InteropLibrary iop)
+        throws UnsupportedMessageException {
+      if (receiver.sealed) {
+        throw UnsupportedMessageException.create();
+      }
+      if (!iop.isDate(value)) {
+        throw UnsupportedMessageException.create();
```

Comment:
This should clearly be `UnsupportedTypeException`.

Reply:
Agreed